test/box-luatest/gh_9823_net_replicaset_test.lua

```diff
@@ -0,0 +1,218 @@
+local t = require('luatest')
+local cbuilder = require('test.config-luatest.cbuilder')
+local cluster = require('test.config-luatest.cluster')
+
+local g = t.group()
+
+g.before_all(cluster.init)
+g.after_all(cluster.clean)
+g.after_each(cluster.drop)
+
+local REDUNDANCY_FACTOR = 3
+
+local function test_config()
+    local sharding_role = {
+        privileges = {{permissions = {'execute'}, universe = true}},
+    }
+    local builder = cbuilder.new()
+    :set_global_option('credentials.roles.sharding', sharding_role)
+    :set_global_option('credentials.users.storage.roles', {'sharding'})
+    :set_global_option('credentials.users.storage.password', 'secret_storage')
+    :set_global_option('iproto.advertise.sharding.login', 'storage')
+    :use_group('test')
+    builder:use_replicaset('router'):add_instance('router', {})
+    builder:use_replicaset('storage')
+           :set_replicaset_option('replication.failover', 'election')
+    for i = 1, REDUNDANCY_FACTOR do
+        builder:add_instance('storage' .. i, {})
+    end
+    return builder:config()
+end
+
+-- Get a valid storage instance name different from given.
+local function get_another_storage_name(storage_name)
+    local num = tonumber(string.sub(storage_name, 8))
+    num = num + 1
+    if num > REDUNDANCY_FACTOR then
+        num = 1
+    end
+    return string.sub(storage_name, 1, 7) .. num
+end
+
+local function get_connect_cfg(cluster)
+    local connect_cfg = {
+        name = 'storage',
+        instances = {},
+    }
+    for i = 1, REDUNDANCY_FACTOR do
+        local instance_name = 'storage' .. i
+        local uri = cluster[instance_name].net_box_uri
+        local instance_config = {
+            endpoint = {
+                uri = uri,
+                login = 'storage',
+                password = 'secret_storage',
+            }
+        }
+        connect_cfg.instances[instance_name] = instance_config
+    end
+    return connect_cfg
+end
+
+-- Test errors of replicaset connector.
+g.test_net_replicaset_error = function(cg)
+    local config = test_config()
+    config.groups.test.replicasets.storage.replication.failover = 'supervised'
+    local cluster = cluster.new(cg, config)
+    cluster:start()
+    local net_replicaset = require('internal.net.replicaset')
+    local connect_cfg = get_connect_cfg(cluster)
+
+    -- Check that unknown replicaset does not work.
+    cluster.router:exec(function()
+        local net_replicaset = require('internal.net.replicaset')
+        local err_msg = 'The replicaset was not found by its name'
+        local succ, err = pcall(net_replicaset.connect, 'moon')
+        t.assert_equals(succ, false)
+        t.assert_equals(err.message, err_msg)
+        t.assert_equals(err.replicaset, 'moon')
+    end)
```

Comment:
It seems you can do it right in the test instance and drop router now completely. This is the only place where router still used. Is there any reason this test case couldn't be done in the main process?

Reply:
Actually there's another place in test_net_replicaset_basics. These tests check connect by replicaset name, and that requires cluster config. The router has has cluster config while the main instance does not, so connection by name have to be tested on router instance.

Reply:
Sad that the config is created in the main process, but somehow can't be used in it.